apps/aecore/lib/aecore/persistence/worker.ex

```diff
@@ -7,50 +7,124 @@ defmodule Aecore.Persistence.Worker do
+
+  @spec get_all_blocks() ::
+  {:ok, map()} | :not_found | {:error, reason :: term()}
+  def get_all_blocks() do
+    GenServer.call(__MODULE__, :get_all_blocks)
+  end
+
+  @spec get_latest_block_height_and_hash() ::
+  {:ok, map()} | :not_found | {:error, reason :: term()}
+  def get_latest_block_height_and_hash() do
+    GenServer.call(__MODULE__, :get_latest_block_height_and_hash)
+  end
+
+  @spec add_account_chain_state(account :: binary(), chain_state :: map()) ::
+  :ok | {:error, reason :: term()}
+  def add_account_chain_state(account, data) do
+    GenServer.call(__MODULE__, {:add_account_chain_state, {account, data}})
   end
-  def write_block_by_hash(_block), do: {:error, "bad block structure"}
 
-  @spec read_block_by_hash(String.t()) ::
-  {:ok, block :: map()} | :not_found | {:error, reason :: term()}
-  def read_block_by_hash(hash) when is_binary(hash) do
-    GenServer.call(__MODULE__, {:read_block_by_hash, hash})
+  @spec get_account_chain_state(account :: binary()) ::
+  {:ok, chain_state :: map()} | :not_found | {:error, reason :: term()}
+  def get_account_chain_state(account) do
+    GenServer.call(__MODULE__, {:get_account_chain_state, account})
+  end
+
+  @spec get_all_accounts_chain_states() ::
+  {:ok, chain_state :: map()} | :not_found | {:error, reason :: term()}
+  def get_all_accounts_chain_states() do
+    GenServer.call(__MODULE__, :get_all_accounts_chain_states)
   end
-  def read_block_by_hash(_hash), do: {:error, "bad hash value"}
 
   ## Server side
 
   def init(_) do
-    ## We are creating `blocks_by_hash` family for the blocks
-    ## https://github.com/facebook/rocksdb/wiki/Column-Families
-    {:ok, db, %{"blocks_by_hash" => blocks_family}} =
+    ## We are ensuring that families for the blocks and chain state
+    ## are created. More about them - https://github.com/facebook/rocksdb/wiki/Column-Families
+    {:ok, db, %{"blocks_family"       => blocks_family,
+                "latest_block_family" => latest_block_family,
+                "chain_state_family"  => chain_state_family}} =
       Rox.open(persistence_path(),
-        [create_if_missing: true, auto_create_column_families: true], ["blocks_by_hash"])
-    {:ok, %{db: db, blocks_family: blocks_family}}
+        [create_if_missing: true, auto_create_column_families: true],
+        ["blocks_family", "latest_block_family", "chain_state_family"])
+    {:ok, %{db: db,
+            blocks_family: blocks_family,
+            latest_block_family: latest_block_family,
+            chain_state_family: chain_state_family}}
   end
 
-  def handle_call({:write_block_by_hash, {hash, block}}, _from,
-    %{db: _db, blocks_family: blocks_family} = state) do
-    {:reply, Rox.put(blocks_family, hash, block), state}
+  def handle_call({:add_block_by_hash, {hash, block}}, _from,
+    %{db: _db, latest_block_family: latest_block_family,
+      blocks_family: blocks_family} = state) do
+    :ok = Rox.put(latest_block_family, @latest_block_key,
```

Comment:
I'm not sure about the purpose of the "letest_block", I guess the idea is to store the top block separately so we can easily find it (without checking all blocks). If so then this put should be corrected (height check should be added), because theoretically we can learn about some shorter fork after learning the current top block.

Reply:
Thank you, for your comment. I will consider this.